docs/rfcs/029-getpage-throttling.md

```diff
@@ -0,0 +1,181 @@
+
+## Solution: Throttling GetPage Requests
+
+**The consequence of the above analysis must be that Pageserver throttles GetPage@LSN requests**.
+That is, unless we want to start charging for provisioned GetPage@LSN/second.
+Throttling sets the correct incentive for a thrashing Compute to scale up its DRAM to the working set size.
+Neon Autoscaling will make this easy, [eventually](https://github.com/neondatabase/neon/pull/3913).
+
+## The Design Space
+
+What that remains is the question about *policy* and *mechanism*:
+
+**Policy** concerns itself with the question of what limit applies to a given connection|timeline|tenant.
+Candidates are:
+
+* hard limit, same limit value per connection|timeline|tenant
+    * Per-tenant will provide an upper bound for the impact of a tenant on a given Pageserver instance.
+      This is a major operational pain point / risk right now.
+* hard limit, configurable per connection|timeline|tenant
+    * This outsources policy to console/control plane, with obvious advantages for flexible structuring of what service we offer to customers.
+    * Note that this is not a mechanism to guarantee a minium provisioned rate, i.e., this is not a mechanism to guarantee a certain QoS for a tenant.
+* fair share among active connections|timelines|tenants per instance
+    * example: divide available IOPS evenly among the tenants that were active in the last 5min
+    * needs definition of "active", and knowledge of available IOPS in advance
+* ...
+
+
+Regarding **mechanism**, it's clear that **backpressure** is the way to go.
+However, we must choose between
+* **implicit** backpressure through pq/TCP and
+* **explicit** rejection of requests + retries with exponential backoff
+
+Further, there is the question of how throttling GetPage@LSN will affect the **internal GetPage latency SLO**:
+where do we measure the SLI for Pageserver's internal getpage latency SLO? Before or after the throttling?
+
+And when we eventually move the measurement point into the Computes (to avoid coordinated omission),
+how do we avoid counting throttling-induced latency toward the internal getpage latency SLI/SLO?
+
+## Scope Of This RFC
+
+**This RFC proposes introducing a hard GetPage@LSN/second limit per tenant, with the same value applying to each tenant on a Pageserver**.
+
+This proposal is easy to implement and significantly de-risks operating large Pageservers,
+based on the assumption that extremely-high-GetPage-rate-episodes like the one from the "Motivation" section are uncorrelated between tenants.
+
+For example, suppose we pick a limit that allows up to 10 tenants to go at limit rate.
+Suppose our Pageserver can serve 100k GetPage/second total at a 100% page cache miss rate.
+If each tenant gets a hard limit of 10k GetPage/second, we can serve up to 10 tenants at limit speed without latency degradation.
```

Comment:
In general I agree with keeping this simple: if I was going to make it just a bit more complicated, here's the permutation that I'd think about:

With sharding, we will ultimately give tenants more op/s as their databases grow in capacity (bigger databases have more shards).  Maybe we should generalize that and define our service to customers as an "IOPs per GB" number (with some clamped minimum number of op/s for small databases and clamped maximum at whatever %ge of the disk max we choose).

The advantage to us is that on a pageserver with many small tenants, this much reduces the probability that several tenants will max out the node: because while those small tenants are numerous, they would also have quite low op/s caps (and, the larger capacity tenants are paying us more).  It also aligns us with the hardware, where EC2 ephemeral storage gives roughly linear IOP/s increase on larger volumes.

It's conceptually very simple, but a little more awkward to implement because it means updating our throttling state at runtime to change the limit (but that's a useful capability anyway).  The configuration/policy would still be global (defining the op/s per GB), I definitely don't want to get into the business of tweaking this tenant-by-tenant.

Reply:
I like that proposal. From a product perspective, I guess we'd use per-GB-of-logicalsize. But, what we really want to limit is the amount of overall work we spend per tenant. The best proxy for that is GB-of-remote-or-local-physical-size-plus-layer-stack-depth or something like that. 🤔

Can we perhaps scope this RFC to use a fixed size, but, require that the implementation allow for dynamic per-tenant reconfiguration of the limit, e.g., derived from tenant size.

Reply:
>Can we perhaps scope this RFC to use a fixed size, but, require that the implementation allow for dynamic per-tenant reconfiguration of the limit, e.g., derived from tenant size.

Yep, makes sense